pandasql/cost_model.py

```diff
@@ -37,29 +39,42 @@ def _out_of_memory(df, graph):
 
 def _join_then_restrict(df, graph):
     for node in graph.keys():
+        # If result is already cached, no point in looking at it
+        if node.result is not None:
+            continue
+
         # TODO: Look at size of filter/limit for this decision?
         if isinstance(node, ps.core.Selection) or \
                 isinstance(node, ps.core.Limit):
-            join_ancestors = _filter_ancestors(node, graph,
-                                               lambda x:
+            join_ancestors = _filter_ancestors(node, graph, lambda x:
                                                isinstance(x, ps.core.Join))
             if len(join_ancestors) > 0:
                 return True
 
-    return False
-
 
 def _limit_output(df, graph):
-    return isinstance(df, ps.core.Limit)
+    if isinstance(df, ps.core.Limit):
+        return True
 
 
 def _deep_dependency_graph(df, graph):
     ancestors_by_depth = _get_ancestors_by_depth(df, graph)
     depth = len(ancestors_by_depth)
-    return depth > 5
+    if depth > 5:
+        return True
+
+
+def _fallback_operation(df, graph):
+    # Rule does not apply if the result of fallback is already cached
+    fallbacks = _filter_ancestors(df, graph, lambda x: x.result is None and
```

Comment:
In a case where the last operation (say a df.describe) cannot be done in current sqlite translation we always run in pandas right now. Could/should we rerun the offload engine on every recursive compute() call to allow offloading for most of a query?

Reply:
Oh interesting. We should look into this when we look into offloading partial queries. I'll put a TODO.